File: app/src/main/java/com/pij/zworkout/service/android/ServiceMappingModule.java
package com.pij.zworkout.service.android;

import com.pij.zworkout.service.api.StorageService;
import com.pij.zworkout.service.api.WorkoutFile;

import java.io.OutputStream;
import java.util.List;

import dagger.Module;
import dagger.Provides;
import dagger.Reusable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;


/**
 * Uses the service module to provide service instance. These are fully working, threaded service implementations directly usable by th app
 * and it presenters and business layer.
 * <p>Created on 04/03/2018.</p>
 *
 * @author Pierrejean
 */
@Module(includes = ServiceModule.class)
public class ServiceMappingModule {

    @Provides
    @Reusable
    Threader provideAndroidThreader() {
        return new AndroidThreader(Schedulers.io(), AndroidSchedulers.mainThread());
    }

    @Provides
    StorageService provideStorageService(FolderStorageService functional, Threader threader) {
        return new StorageService() {
            @Override
            public Observable<List<WorkoutFile>> workouts() {
                return functional.workouts().compose(threader.forObservable());
            }

            @Override
            public Single<OutputStream> openForWrite(WorkoutFile file) {
                return functional.openForWrite(file).compose(threader.forSingle());
            }
        };
    }

}

File: app/src/main/java/com/pij/zworkout/service/android/ServiceMappingModule.kt
/*
 * Copyright (c) 2018, Chiswick Forest
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and limitations under the License.
 */

package com.pij.zworkout.service.android

import com.pij.zworkout.service.api.StorageService
import dagger.Module
import dagger.Provides
import dagger.Reusable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.io.File


/**
 * Uses the service module to provide service instance. These are fully working, threaded service implementations directly usable by th app
 * and it presenters and business layer.
 *
 * Created on 04/03/2018.
 *
 * @author Pierrejean
 */
@Module(includes = [(ServiceModule::class)])
class ServiceMappingModule {

    @Provides
    @Reusable
    internal fun provideAndroidThreader(): Threader {
        return AndroidThreader(Schedulers.io(), AndroidSchedulers.mainThread())
    }

    @Provides
    internal fun provideStorageService(functional: FolderStorageService, threader: Threader): StorageService {
        return object : StorageService {
            override fun workouts() = functional.workouts().compose(threader.forSingle())

            override fun create(name: String) = functional.create(name).compose(threader.forSingle())

            override fun openForWrite(target: File) = functional.openForWrite(target).compose(threader.forSingle())

            override fun openForRead(source: File) = functional.openForRead(source).compose(threader.forSingle())
        }
    }

}
